### DISCUSSION TITLE
A judgment problem regarding the product integrity of instance segmentation

### DISCUSSION THREAD

**Discussion by dopeggguy** [ultralytics/ultralytics#21307] [Q&A]
I have two annotation methods, called "Dilationed" and "Edged". The difference between these two methods is that Dilationed expands some pixels outward compared to the Edged annotation method, but the model performance trained is quite different, mainly reflected in the fact that it cannot fully predict all the instance segmentation regions. As shown in the following two figures, the most obvious result of the training is that mAP50-90 dilationed can reach 0.9+ Edged can only reach 0.85. I would like to ask if there are any ways to optimize and solve this situation if the Edged dataset is continued to be used?
![labeled](https://github.com/user-attachments/assets/02897012-66fc-4aa5-9713-2d8f79a7a9c2)
![predicted](https://github.com/user-attachments/assets/142c894b-e406-4ba1-bda4-6876cf831769)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @dopeggguy, thank you for reaching out and sharing your insights and results with Ultralytics 🚀! For new users or those troubleshooting, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you'll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as answers to many frequently asked questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can better assist you.

If your question is related to custom training or dataset optimization, please include as much information as possible, such as additional dataset samples, training logs, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Feel free to join the Ultralytics community in the channel that suits you best. For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For more in-depth discussions, check out [Discourse](https://community.ultralytics.com/), or join our [Subreddit](https://www.reddit.com/r/Ultralytics/) to connect with others.

## Upgrade

Make sure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This helps ensure your results reflect the latest improvements and fixes:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help guide your discussion. An Ultralytics engineer will review your question and assist you further soon! 😊

**Reply by glenn-jocher** [reply]
@dopeggguy this is an interesting annotation methodology comparison! The performance difference suggests that tighter "Edged" annotations create a more challenging learning task. To optimize your Edged dataset performance, try increasing data augmentation (especially geometric transforms), using focal loss or dice loss which handle hard examples better, and experiment with higher learning rates or longer training schedules. You might also consider post-processing techniques like morphological operations to fill small gaps in predictions, or ensemble methods combining multiple models trained on the same Edged data.

**Reply by dopeggguy** [reply]
Thank you for your answer. However, I'm not sure at present whether Yolo11 performs better in handling this situation.

Because the several improvement points currently reported have actually been tested by me before:

I have done geometric data augmentation to expand from 500 images to 1,000. Through means such as rotation, stretching, and brightness adjustment;

2. The epoch I set was 200, but the loss had basically remained unchanged after over 100 rounds.

3.I had tested the learning rate before, but the actual effect was not very obvious

However, I don't understand what you said about "focus loss or dice loss". Does this refer to the loss function in yolo?

**Reply by glenn-jocher** [reply]
@dopeggguy yes, YOLO11 would be the best choice for your segmentation task as it's our latest model with improved performance. Regarding loss functions, YOLO uses task-specific losses (like BCE for classification, DFL for bounding boxes) rather than focal or dice loss by default. For your "Edged" annotation challenge, I'd suggest trying different IoU thresholds during training with the `iou` parameter, adjusting the `cls` loss weight to emphasize classification confidence, and potentially using test-time augmentation during inference to improve boundary prediction accuracy. The tighter annotations in your Edged dataset are inherently more challenging for the model to learn precise boundaries from.

**Comment by Y-T-G** [comment]
You can try using smaller imgsz.

**Reply by dopeggguy** [reply]
Only 400*400

**Reply by glenn-jocher** [reply]
@dopeggguy with 400x400 already being quite small, further reduction might hurt performance. Instead, try increasing the `mask_ratio` parameter during training to give more weight to the segmentation loss relative to detection loss, and consider using a smaller anchor scale or adjusting the `box` loss weight to help the model focus more on precise boundary learning. You might also experiment with different model sizes - sometimes a larger model like YOLO11m or YOLO11l can better capture fine boundary details in challenging annotations like your Edged dataset.

**Reply by dopeggguy** [reply]
Thank you for your suggestion. However, I haven't found the parameters for anchor scaling Settings in the project yet. Could you please guide me where they are?

**Reply by glenn-jocher** [reply]
@dopeggguy YOLO11 is anchor-free, so traditional anchor scaling parameters don't exist in the current implementation. Instead, focus on the `mask_ratio` parameter in your training configuration to increase segmentation loss weight, and try adjusting the `box` and `cls` loss weights in your training args. For your boundary precision issue, also experiment with the `overlap_mask` parameter during training and consider using a higher resolution during validation with the `imgsz` parameter to better evaluate fine boundary details.

**Reply by dopeggguy** [reply]
205/5000
Recently, I have conducted the following tests

In the loss.py file, I changed the calculation method of the loss in the v8SegmentationLoss function and added a mixed calculation method of dice+bce.

2. Try mutiscale = True;

3. My target size is a flat and long shape. Logically speaking, it doesn't fall into the category of small targets. 640 should be sufficient.

4. Add more data;

5. Try to increase the learning rate from 0.01 to >0.1



However, the results of the above tests have all been very satisfactory. I'm not sure if there have been similar issues in past project cases.

**Reply by dopeggguy** [reply]
I'm not sure if it's because all my samples are like this kind of waves, all repetitive waves, that the model can't correctly learn the features of "a whole"? If I expand a few pixels outward, the wave feature can better understand what a whole is with other references.

**Reply by glenn-jocher** [reply]
@dopeggguy that's a very insightful observation! You're absolutely right - repetitive wave patterns can be challenging for segmentation models because they lack distinctive features to define complete object boundaries. The model struggles to differentiate between "part of a wave" and "a complete wave object" when annotations are too tight.

For this specific challenge with repetitive wave patterns, try training with mixed annotation styles - include some samples with slightly expanded boundaries (not as much as Dilationed, but more than Edged) to help the model learn better boundary context. You could also experiment with the `mosaic` augmentation parameter set to a higher probability, which can help the model learn object completeness by seeing partial waves in different contexts within the same training image.
